Src/Support/Google.Apis.Auth/OAuth2/GoogleCredential.cs

```diff
@@ -179,32 +182,49 @@ public static async Task<GoogleCredential> FromFileAsync(string path, Cancellati
         /// <returns>A credential based on the provided access token.</returns>
         public static GoogleCredential FromAccessToken(string accessToken, IAccessMethod accessMethod = null)
         {
-            accessMethod = accessMethod ?? new BearerToken.AuthorizationHeaderAccessMethod();
+            accessMethod ??= new BearerToken.AuthorizationHeaderAccessMethod();
             return new GoogleCredential(new AccessTokenCredential(accessToken, accessMethod));
         }
 
-        internal class AccessTokenCredential : ICredential, IHttpExecuteInterceptor
+        internal class AccessTokenCredential : IGoogleCredential, IHttpExecuteInterceptor
         {
-            public AccessTokenCredential(string accessToken, IAccessMethod accessMethod)
+            private readonly string _accessToken;
+            private readonly IAccessMethod _accessMethod;
+
+            /// <inheritdoc/>
+            public string QuotaProject { get; }
+
+            public AccessTokenCredential(string accessToken, IAccessMethod accessMethod, string quotaProject = null)
             {
                 _accessToken = accessToken;
                 _accessMethod = accessMethod;
+                QuotaProject = quotaProject;
             }
 
-            private readonly string _accessToken;
-            private readonly IAccessMethod _accessMethod;
+            /// <inheritdoc/>
+            IGoogleCredential IGoogleCredential.WithQuotaProject(string quotaProject) =>
+                new AccessTokenCredential(_accessToken, _accessMethod, quotaProject);
 
             public void Initialize(ConfigurableHttpClient httpClient) =>
                 httpClient.MessageHandler.Credential = this;
 
             public Task<string> GetAccessTokenForRequestAsync(string authUri = null, CancellationToken cancellationToken = default(CancellationToken)) =>
                 Task.FromResult(_accessToken);
 
+            /// <inheritdoc />
+            public Task<AccessTokenWithHeaders> GetAccessTokenWithHeadersForRequestAsync(string authUri = null, CancellationToken cancellationToken = default) =>
+                Task.FromResult(new AccessTokenWithHeaders(_accessToken, QuotaProject));
+
             // Only method in IHttpExecuteInterceptor
-            public Task InterceptAsync(HttpRequestMessage request, CancellationToken taskCancellationToken)
+            public async Task InterceptAsync(HttpRequestMessage request, CancellationToken taskCancellationToken)
```

Comment:
This would be more efficient with a *little* code duplication:

```csharp
public Task InterceptAsync(HttpRequestMessage request, CancellationToken taskCancellationToken)
{
    var accessToken = new AccessTokenWithHeaders(_accessToken, QuotaProject);
    _accessMethod.Intercept(request, accessToken.AccessToken);
    foreach (var header in accessToken.Headers)
    {
        request.Headers.Add(header.Key, header.Value);
    }
    return Task.FromResult(0);
}
```

There's no need to construct all the async machinery.

Reply:
Yes, done.